Sources/Brave/Frontend/Browser/Search/Voice Search/SpeechRecognizer.swift

```diff
@@ -0,0 +1,300 @@
+        guard await SFSpeechRecognizer.hasAuthorizationToRecognize() else {
+          throw RecognizerError.authorizationAccessDenied
+        }
+      }
+      
+      return true
+    } catch {
+      log.debug("Voice Search Authorization Fault \(error.localizedDescription)")
+    }
+    
+    return false
+  }
+  
+  @MainActor
+  func startTranscribing() {
+    transcribe()
+  }
+    
+  @MainActor
+  func stopTranscribing() {
+    reset()
+  }
+  
+  func startSilenceAnimation() {
+    animationType = .pulse(scale: AnimationScale.pulse)
+  }
+
+  func startSpeechAnimation(_ scale: CGFloat) {
+    animationType = .speech(volume: scale)
+  }
+  
+  /// Creates a `SFSpeechRecognitionTask` that transcribes speech to text until you call `stopTranscribing()`.
+  /// The resulting transcription is continuously written to the published `transcript` property.
+  private func transcribe() {
+    guard let recognizer, recognizer.isAvailable else {
+      log.debug("Voice Search Unavailable: ")
+      return
+    }
+    
+    do {
+      let (audioEngine, request) = try setupStartEngine()
+        
+      self.audioEngine = audioEngine
+      self.request = request
+      
+      task = recognizer.recognitionTask(with: request, resultHandler: { [weak self] result, error in
+        guard let self = self else {
+          return
+        }
+        
+        var isFinal = false
+        
+        if let result, !isFinal {
+          let formattedTranscript = result.bestTranscription.formattedString
+          let transcriptComponents = formattedTranscript.components(separatedBy: .whitespacesAndNewlines)
+          let formattedWords = transcriptComponents.filter { !$0.isEmpty }
+          
+          // SpeechRecognitionMetadata is the key to detect speaking finalized
+          isFinal = result.isFinal
+          || result.speechRecognitionMetadata != nil
+          || formattedWords.count >= 15
+              
+          if !formattedTranscript.isEmpty {
+            self.transcribe(formattedTranscript)
+          }
+        }
+        
+        // Check voice input final
+        if isFinal {
+          animationType = .stable
+          transcriptedIcon = "leo.check.circle-outline"
+          
+          // Remove audio buffer input
+          audioEngine.inputNode.removeTap(onBus: 0)
+          // Reset Speech Recognizer
+          self.reset()
+
+          Task.delayed(bySeconds: 0.75) { @MainActor in
+            if let formattedTranscript = result?.bestTranscription.formattedString, !formattedTranscript.isEmpty {
+              self.finalize(searchQuery: formattedTranscript)
+            }
+          }
+        }
+      })
+    } catch {
+      reset()
+      log.debug("Voice Search Recognization Fault \(error.localizedDescription)")
+    }
+  }
+    
+  /// Reset the speech recognizer.
+  private func reset() {
+    try? AVAudioSession.sharedInstance().setActive(false)
```

Comment:
What happens if I have playlist playing in the background and voice search gets reset? I believe this line will completely disable playlist.

Reply:
- YouTube playing in the background then do a voice search and lock the screen.

So for this one the moment you pressed Voice Search button youtube pauses and when search finished (done - canceled) the youtube continues automatically


- Pip enabled and voice search, does it disable sound? If so, what happens if I can the search? Does the sound come back?

Yes it disables the sound and pauses the video and when search is done, video is still paused but can press continue

Reply:
A note for QA to test with Background audio and playlist is included inside description.

Reply:
Made some further fixes https://github.com/brave/brave-ios/pull/7551/commits/ccddc2a772fc2cc8049b706b9fa1f3fc91276b09